rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,1222 @@
+
+- Script-Based Installs MVP: Early MVP supporting only manual control and
+simple script-based installers for non-auth instances.
+
+- TUF-Based System MVP: Fully functional but minimalistic upgrade system
+based on TUF (still excludes auth instances).
+
+- Stability & Polish: Additional necessary features (including auth installs
+and local rollbacks). Represents the point at which the core upgrade system
+can be considered "complete".
+
+- Extended Feature Set: A collection of nice-to-haves that we're pretty sure
+folks are going to want.
+
+See the [implementation plan](#implementation-plan) section for detailed breakdown
+of the required elements for each phase.
+
+
+## Details
+
+### Usage Scenarios
+
+
+#### Notification-Only Usecase
+
+Cluster administrators may or may not be using teleport's own install mechanisms, but they
+want teleport to be able to inform them when instances are outdated, and possibly generate
+alerts if teleport is running a deprecated version and/or there is a newer security patch
+available.
+
+In this case we want options for both displaying suggested version (or just a "needs upgrade"
+badge) on inventory views, and also probably some kind of cluster-wide alert that can be
+made difficult to ignore (e.g. a message on login or a banner in the web UI). We also
+probably want an API that will support plugins that can emit alerts to external locations
+(e.g. a slack channel).
+
+In this usecase teleport is serving up recommendations based on external state, so a client
+capable of discovery (e.g. the `tuf` version controller) is required, but the actual
+ability to affect installations may not be necessary.
+
+
+#### Minimal Installs Usecase
+
+Cluster administrators explicitly set the desired version(s) of teleport installations, and
+have a specific in-house installation script that should be used for upgrades. Teleport cluster
+may not even have any access to the public internet.  The installation target may be an internally
+built fork of teleport.
+
+In this case, we want to provide the means to specify the target version and desired script. Teleport
+should then be able to detect when a teleport node is not running the target that it aught to, and
+invoke the upgrade script.  Teleport should not care about the internals of how the script plans to
+perform the installation.  Nodes that require upgrades run the script, and may be required to perform
+a graceful restart if the script succeeds. The script may expect inputs (e.g. version string), and
+there may be different scripts to run depending on the nature of the specific node (e.g. `prod-upgrade`, `staging-upgrade`
+or `apt-upgrade`, `yum-upgrade`), but things like selecting the correct processor architecture or OS
+compatibility are likely handled by the script.
+
+In this minimal usecase teleport's role is primarily as a coordinator.  It detects when and where user
+provided scripts should be run, and invokes them.
```

Comment:
What happens between the time that teleport invokes the script and the time the upgrade completes? 

- Does teleport somehow know/track that it has an upgrade in progress?
- Is there a period after which time we consider the upgrade to have failed?
- What if the script fails? Do we retry? With any type of backoff?